src/librustc_mir/borrow_check.rs

```diff
@@ -446,54 +446,78 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
+                            let end_issued_loan_span =
+                                flow_state.borrows.base_results
+                                    .operator()
+                                    .opt_region_end_span(&borrow.region);
+                            this.report_conflicting_borrow(
+                                context, common_prefix, lvalue_span, bk,
+                                &borrow, end_issued_loan_span)
                         }
-                        Control::Break
                     }
-                    (Write(kind), _) => {
-                        match kind {
-                            WriteKind::MutableBorrow(bk) => {
-                                let end_issued_loan_span =
-                                    flow_state.borrows.base_results.operator().opt_region_end_span(
-                                        &borrow.region);
-                                this.report_conflicting_borrow(
-                                    context, common_prefix, lvalue_span, bk,
-                                    &borrow, end_issued_loan_span)
-                            }
-                            WriteKind::StorageDead |
-                            WriteKind::Mutate =>
-                                this.report_illegal_mutation_of_borrowed(
-                                    context, lvalue_span, borrow),
-                            WriteKind::Move =>
-                                this.report_move_out_while_borrowed(
-                                    context, lvalue_span, &borrow),
+                    Control::Break
+                }
+                (Write(kind), _) => {
+                    match kind {
+                        WriteKind::MutableBorrow(bk) => {
+                            let end_issued_loan_span =
+                                flow_state.borrows.base_results
+                                    .operator()
+                                    .opt_region_end_span(&borrow.region);
+                            this.report_conflicting_borrow(
+                                context, common_prefix, lvalue_span, bk,
+                                &borrow, end_issued_loan_span)
                         }
-                        Control::Break
+                        WriteKind::StorageDead |
+                        WriteKind::Mutate =>
+                            this.report_illegal_mutation_of_borrowed(
+                                context, lvalue_span, borrow),
+                        WriteKind::Move =>
+                            this.report_move_out_while_borrowed(
+                                context, lvalue_span, &borrow),
                     }
+                    Control::Break
                 }
-            });
+            }
+        }
+
+        // Check permissions
+        self.check_access_permissions(lvalue_span, rw);
+
+        match *self.prefixes(lvalue_span.0, PrefixSet::All).last().unwrap() {
```

Comment:
Shouldn't borrows involving statics be handled in the same way as borrows involving raw pointers? I could open another issue for borrows through raw pointers.

```Rust
struct Foo { x: [usize; 2] }

static mut SFOO: Foo = Foo { x: [23, 32] }; 

impl Foo {
    fn x(&mut self) -> &mut usize { &mut self.x[0] }
}

fn main() {
    // this compiles with AST borrowck, should also compile with MIR borrowck
    unsafe {
        let sfoo: *mut Foo = &mut SFOO;
        let x = (*sfoo).x();
        (*sfoo).x[1] += 1;
        *x += 1; 
    }
}
```

Reply:
Yes, I agree they should be handled the same. Do you think that means however that we should unify the codepaths that make those two cases behave the same? If so, how do you think we should do it? (I think that sounds like a nice idea)

Reply:
(In particular, if we decided to offer a lint at some point for obvious UB, I'd want it to apply to both.)

Reply:
I guess if we just factor out a helper function `unsafe_access_path(&Lvalue)` it can test for either a deref of a raw pointer *or* the use of a `static mut`.

Reply:
@nikomatsakis 

One I think good way of doing it is just not registering borrows that go through an unsafe access path.

Note that for unsafe pointers this *is* different from checking the lvalue, e.g. in:
```Rust
fn main() {
    let mut x = 1 as *const ();
    let y = unsafe { &*x }; // borrow through an unsafe pointer
    let _ = &mut x; // then mutate everything - should be OK
    println!("{:p}", y);
}
```

Reply:
I was avoiding recommending that because I wanted to consider adding a lint later on. But we could do it that way. Either way, we basically need the helper function I talked about, I imagine. It's just a question of whether we try to avoid registering the borrows in the first place, or try to ignore them when they are registered. =)